File: core/src/main/java/com/codeabovelab/tpc/core/processor/Rule.java
package com.codeabovelab.tpc.core.processor;

import lombok.Data;

/**
 */
@Data
public final class Rule implements Comparable<Rule> {

    private final String id;
    private final float weight;
    private final RulePredicate predicate;
    private final RuleAction action;


    @Override
    public int compareTo(Rule o) {
        return Float.compare(this.getWeight(), o.getWeight());
    }
}

File: core/src/main/java/com/codeabovelab/tpc/core/processor/Rule.kt
package com.codeabovelab.tpc.core.processor

/**
 */
class Rule<T: PredicateResult<*>>(
    /**
     * Id of rule. It must be unique for whole system.
     */
    val id: String,
    /**
     * Used for order of rules before applying to document. Less weight rules apply first.
     */
    val weight: Float,
    /**
     * Predicate which determine of applicability of rule.
     */
    val predicate: RulePredicate<T>,
    val action: RuleAction<T> = RuleAction.NOP
): Comparable<Rule<*>> {

    override fun compareTo(other: Rule<*>): Int {
        return this.weight.compareTo(other.weight)
    }
}
